compiler/src/dotty/tools/dotc/core/SymDenotations.scala

```diff
@@ -2351,8 +2351,11 @@ object SymDenotations {
     if (denot.isOneOf(ValidForeverFlags) || denot.isRefinementClass || denot.isImport) true
     else {
       val initial = denot.initial
-      val firstPhaseId = initial.validFor.firstPhaseId.max(typerPhase.id)
-      if ((initial ne denot) || ctx.phaseId != firstPhaseId)
+      val firstPhaseId =
+        initial.validFor.firstPhaseId.max(typerPhase.id)
+      if firstPhaseId > ctx.lastPhaseId then
+        false
+      else if (initial ne denot) || ctx.phaseId != firstPhaseId then
```

Comment:
If the phases have changed (e.g., adding or removing some phases in the middle), then the periods of denotations will be incorrect in the new run. Therefore, do we need to invalidate all non-permanent denotations in such cases, as the same `phaseId` does not mean the same phase?


https://github.com/lampepfl/dotty/blob/64a239f62ce8a9528b3a4e2c52349180629ef570/compiler/src/dotty/tools/repl/ReplCompiler.scala#L34-L39

Reply:
Yes, that looks like a problem. But can phases change like this in the same compiler instance? In that case I don't see how we can re-use any symbols. So it might be better to just start a new compiler instance in that case.